### ISSUE TITLE
Enhancement: Allow setting OS Name and Version for a computer

### ISSUE THREAD

**Issue by oflebbe** [msktutil/msktutil#26]
*Originally created by:* [fran.fav...@gmail.com](http://code.google.com/u/117366488858131274109/)

Similar to --description, adding options to set the OS name \(--osname\) and version \(--osver\) for a computer would be useful. These are the operatingSystem and operatingSystemVersion LDAP attributes. Perhaps they could default to uname -s and uname -r. Samba has similar options for net ads join.

For completeness, you could add OS service pack \(--ossp\) defaulting to uname -v.

Reported by: *anonymous

**Comment by michael-o** [comment]
I just tried this local patch:

    diff --git a/msktldap.cpp b/msktldap.cpp
    index 51c66ec..53df13d 100644
    --- a/msktldap.cpp
    +++ b/msktldap.cpp
    @@ -30,6 +30,7 @@
      */
    
     #include "msktutil.h"
    +#include <sys/utsname.h>
     #include <algorithm>
     #include <sstream>
     #include <iostream>
    @@ -433,6 +434,19 @@ void ldap_check_account_strings(msktutil_flags *flags)
             }
         }
    
    +    if (!flags->use_service_account) {
    +        struct utsname system_info;
    +        int ret = uname(&system_info);
    +        if (!ret) {
    +            std::string sysname = std::string(system_info.sysname);
    +            std::string release = std::string(system_info.release);
    +            ldap_simple_set_attr(dn, "operatingSystem", sysname, flags);
    +            ldap_simple_set_attr(dn, "operatingSystemVersion", release, flags);
    +        } else {
    +            /* TODO: Print a message? */
    +        }
    +    }
    +
   
and recieved this message:

    ]$ ./msktutil update --computer-name wrkst --verbose --account-name wrkst\$ --keytab wrkst.keytab
     -- init_password: Wiping the computer password structure
     -- generate_new_password: Generating a new, random password for the computer account
     -- generate_new_password:  Characters read from /dev/urandom = 98
     -- get_dc_host: Attempting to find Domain Controller to use via DNS SRV record in domain AD001.OSIPOV.EU for procotol tcp
     -- get_dc_host: Attempting to find Domain Controller to use via DNS SRV record in domain AD001.OSIPOV.EU for procotol udp
     -- get_dc_host: Attempting to find a Domain Controller to use (DNS domain)
     -- validate: Found DC: AD001.OSIPOV.EU. Checking availability...
     -- get_dc_host: Found preferred Domain Controller: bsd1dc.ad001.osipov.eu
     -- create_fake_krb5_conf: Created a fake krb5.conf file: /tmp/.msktkrb5.conf-xQ2rnA
     -- destroy_g_context: Destroying Kerberos Context
     -- initialize_g_context: Creating Kerberos Context
     -- finalize_exec: SAM Account Name is: wrkst$
     -- try_machine_keytab_princ: Trying to authenticate for wrkst$ from local keytab
     -- switch_default_ccache: Using the local credential cache: FILE:/tmp/.mskt_krb5_ccache-Me11do
     -- finalize_exec: Authenticated using method 1
     -- LDAPConnection: Connecting to LDAP server: bsd1dc.ad001.osipov.eu
    SASL/GSSAPI authentication started
    SASL username: wrkst$@AD001.OSIPOV.EU
    SASL SSF: 56
    SASL data security layer installed.
     -- ldap_get_base_dn: Determining default LDAP base: dc=AD001,dc=OSIPOV,dc=EU
     -- get_default_ou: Determining default OU: CN=Computers,DC=ad001,DC=osipov,DC=eu
     -- ldap_check_account: Checking that a computer account for wrkst$ exists
     -- ldap_check_account: Checking computer account - found
     -- ldap_check_account: Found userAccountControl = 0x1000
     -- ldap_check_account: Found supportedEncryptionTypes = 28
     -- ldap_check_account: Found dNSHostName = bsd1srv.ad001.osipov.eu
     -- ldap_check_account: Found Principal: host/bsd1srv.ad001.osipov.eu
     -- ldap_check_account: Found Principal: host/wrkst
     -- ldap_check_account_strings: Inspecting (and updating) computer account attributes
     -- modify_ext: ldap_modify_ext_s failed (Insufficient access)
    WARNING: ldap modification of CN=wrkst,CN=Computers,DC=ad001,DC=osipov,DC=eu
             failed while trying to change operatingSystem to FreeBSD.
             Error was: Insufficient access
             --> Do you have enough privileges?
             --> You might try re-"kinit"ing.
             --> Maybe you should try again with --user-creds-only?
             Continuing anyway ...
     -- modify_ext: ldap_modify_ext_s failed (Insufficient access)
    WARNING: ldap modification of CN=wrkst,CN=Computers,DC=ad001,DC=osipov,DC=eu
             failed while trying to change operatingSystemVersion to 10.3-RELEASE-p20.
             Error was: Insufficient access
             --> Do you have enough privileges?
             --> You might try re-"kinit"ing.
             --> Maybe you should try again with --user-creds-only?
             Continuing anyway ...
     -- ldap_set_supportedEncryptionTypes: No need to change msDs-supportedEncryptionTypes they are 28
     -- ldap_set_userAccountControl_flag: Setting userAccountControl bit at 0x200000 to 0x0
     -- ldap_set_userAccountControl_flag: userAccountControl not changed 0x1000
     -- ldap_get_kvno: KVNO is 19
    
Does it really require admin privs? I did `kinit` with `Administrator@AD001.OSIPOV.EU`  and used `--user-creds-only` and it worked. That seems pretty annoying if you want to update the version of your kernel or OS. It suits for `create` mode only.

**Comment by michael-o** [comment]
@oflebbe @mproehl any comments?

**Comment by mproehl** [comment]
Yes, It seems to require admin privs:

```
# ldapmodify 
SASL/GSSAPI authentication started
SASL username: lx02$@ADS.EXAMPLE.COM
SASL SSF: 56
SASL data security layer installed.
dn: CN=lx02,CN=Computers,DC=ads,DC=example,DC=com
changetype: modify
add: servicePrincipalName
servicePrincipalName: test/lx02

modifying entry "CN=lx02,CN=Computers,DC=ads,DC=example,DC=com"

dn: CN=lx02,CN=Computers,DC=ads,DC=example,DC=com
changetype: modify
add: operatingSystem
operatingSystem: debian

modifying entry "CN=lx02,CN=Computers,DC=ads,DC=example,DC=com"
ldap_modify: Insufficient access (50)
	additional info: 00002098: SecErr: DSID-03150F93, problem 4003 (INSUFF_ACCESS_RIGHTS), data 0
```

While `servicePrincipalName` can be updated by the machine itself (at lease if the value has the hostname as 2nd component) -- `operatingSystem` can not. We require admin privs.

I would suggest to let `--osname` and `--osver` set `user-creds-only`. What do you think?

When not called with `--osname` or `--osver`, we could add the `utsname`patch but only try to set the according attributes without failing if we do not have admin privs.

**Comment by mproehl** [comment]
On the other hand... Windows computers do not need admin privs to update those attrs ...

**Comment by michael-o** [comment]
@mproehl Any idea why? Don't they do it over LDAP? Does Windows perform any RPC calls here?

**Comment by glorang** [comment]
We've been looking into this as well, looks like Windows clients are using MS-RPC NetrLogonGetDomainInfo() instead of LDAP update.

https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-nrpc/7c3ad0cc-ee05-4643-b773-4d84e1d431dc
https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-nrpc/3ae9e9a9-a303-4fa5-8e11-823d9e7e1e61

I've just posted this on Samba mailing list as well, guess they will now more/better.

**Comment by michael-o** [comment]
@glorang, strange...the procedure says `GetDomainInfo()`. Does this really perform a write operation?

**Comment by glorang** [comment]
Yes, so it seems. See the links I attached. In the NetrLogonGetDomainInfo you add a "NETLOGON_WORKSTATION_INFORMATION" struct: 

> The NETLOGON_WORKSTATION_INFO structure defines information passed into the NetrLogonGetDomainInfo method, as specified in 3.5.4.4.9. It SHOULD<20> be used to convey information about a member workstation from the client side to the server side.

**Comment by michael-o** [comment]
What a horrible API design.

**Comment by michael-o** [comment]
I think this will not happen for 1.2, will it?

**Comment by dkobras** [comment]
Indeed this looks like a simple add-on, but turns out the be quite tricky to implement. Hence I've removed the issue from the v1.2 milestone. Still, we'll have to decide how to pursue here. It seems we can either implement (partial) support for Netlogon, or go for a simler, LDAP-based version that requires admin credentials, or choose not to implement any of this at all.
